0015-estimator-interface.md

```diff
@@ -0,0 +1,500 @@
+For example, all weight-1 and weight-2 Paulis that are adjacent on the connectivity graph. 
+For a one-hundred qubit device, this corresponds to hundreds of unique estimates to be made for a single circuit, noting that for this particular example, on the heavy-hex graph, in the absence of mitigation, only 9 circuits need to be physically run.
+
+We propose a new `ObservablesArray` object, where each element corresponds to an observable the user wants an estimated expectation value of.
+We propose that the internal data model of this object have an element format
+
+```python
+List[Tuple[str, List[int], float]]
+```
+
+where, for example, the observable `0.5 * IIZY + 0.2 * XIII` would be stored as
+
+```python
+[("ZY", [2, 3], 0.5), ("X", [0], 0.2)]
+```
+
+This is proposed instead of `{"IIZY": 0.5, "XIII": 0.2}` anticipating the overhead of storing and working with so many `"I"`'s for large devices.
+
+Note that the above is simply the data model; the `ObservablesArray` object, either in the constructor or special static coercion method, will accept the `quantum_info` objects that are convenient to users.
+However, only types such as `Pauli` and `SparsePauliOp` that are known to have sparse Pauli representations will be allowed; we will drop general support for `BaseOperator`.
+Further, through [type coersion](#type-coercion-strategy) in `Estimator.run()` itself, a user won't need to manually construct an `ObservablesArray`, as shown in the example in the [Design Proposal](#design-proposal) section.
+
+It is up to each `Estimator` implementation to, if it sees fit, to solve a graph coloring problem that decides how to produce a sufficient set of physical circuits that are capable of producing data to make each of these estimates.
+
+### Arrays and Broadcasting <a name="arrays-and-broadcasting"></a>
+
+An `nd-array` is an object whose elements are indexed by a tuple of integers, where each integer must be bounded by the dimension of that axis. 
+The tuple of dimensions, one for each axis, is called the shape of the `nd-array`. 
+For example, 1D list of 10 objects (typically numbers) is a vector and has shape `(10,)`; a matrix with 5 rows and 2 columns is 2D and has shape `(5, 2)`; a single number is 0D and has an empty shape tuple `()`; an `nd-array` with shape (20, 10, 5) can be interpreted as a length-20 list of 10×5 matrices.
+
+We propose the constraint that a `BindingsArray` instance and an `ObservablesArray` instance that live in the same task must have shapes that are broadcastable, in which case, the `Estimator` promises to return one expectation value estimate for each element of the broadcasted shape.
+
+Here are some examples of common patterns expressed in terms of array broadcasting, and their accompanying visual representation in the figure that follows:
+
+```python
+# Broadcast single observable
+parameter_values = np.array(1.0)
+parameter_values.shape == ()
```

Comment:
Can I suggest not allowing the behavior of implicitly appending extra (1) dimensions for broadcasting? I find this aspect of matlab/numpy array-handling to be error-prone and I'd prefer explicit dimensions in all cases, whether prepending (where it is already required) or appending (where it is currently optional):
yes (values.shape, observables.shape -> results.shape):

> (5,) (1,) ->(5,)
> (5,7) (5,1) -> (5,7)
> (5,7) (1,7) -> (5,7)

No:

> (5,) () -> (5,)
> (5,7) (5,) -> (5,7)

Providing input in this form is explicit using np.newaxis, which just fiddles with array metadata doesn't require memcpy or anything:

```
In [9]: x.shape
Out[9]: (2, 3)

In [10]: x[:,np.newaxis,:,np.newaxis].shape
Out[10]: (2, 1, 3, 1)
```

Reply:
That's interesting. By the way, the numpy convention is to right align (I blundered in some version of this document), so `(5,7) (5,) -> (5,7)` should actually be `(5,7) (7,) -> (5,7)`. But this serves your point.

Despite this, I do think there's value in being able to say "it's just like numpy", and I also think there's value in being able to provide 0-D values to broadcast. Maybe others want to weigh in.

Reply:
I propose to hash it out in the PRs.

Reply:
I don't have a strong opinion either way and agree we can hash it out in PRs just wanted to raise the question